<a href="https://colab.research.google.com/github/Pamela2710/Data-Mining/blob/main/DataProcessing/Copia_de_DataProcessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
import pandas as pd
import numpy as np
import random
from sklearn.preprocessing import MinMaxScaler, LabelEncoder
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import SelectFromModel, mutual_info_classif
from sklearn.metrics import accuracy_score, make_scorer
from sklearn.base import clone
from scipy.stats import entropy
from sklearn.datasets import make_classification
from sklearn.feature_selection import SelectKBest

df = pd.read_csv('https://raw.githubusercontent.com/Pamela2710/Data-Mining/main/DataProcessing/features_completas.csv', encoding='latin-1')

# Eliminar filas con valores NaN
df = df.dropna()

df

Media tiempo  Desviacion estandar tiempo  Varianza tiempo  \
0        0.099520                    0.245902         0.108890   
1        0.085819                    0.876726         0.796870   
2        0.077419                    0.231336         0.099950   
3        0.102426                    0.291445         0.138865   
4        0.091737                    0.428982         0.247991   
..            ...                         ...              ...   
685      0.101867                    0.214477         0.089995   
686      0.086898                    0.680620         0.520007   
687      0.077123                    0.548769         0.365809   
688      0.079244                    0.720172         0.571271   
689      0.087387                    0.129870         0.046375   

     Entropia tiempo  kurtosis tiempo  MSE tiempo  \
0                  0         0.013447    0.907095   
1                  0         0.050793    0.351758   
2                  0         0.015484    0.882481   
3                  0         0.011577    0.910015   
4                  0         0.026587    0.723575   
..               ...              ...         ...   
685                0         0.018980    0.875568   
686                0         0.038755    0.298080   
687                0         0.024871    0.207060   
688                0         0.031290    0.165912   
689                0         0.010445    0.986443   

     Tiempo en alcanzar el pico maximo en frecuencia  RMS tiempo  \
0                                           0.115646    0.245920   
1                                           0.306122    0.876697   
2                                           0.122449    0.231357   
3                                           0.102041    0.291448   
4                                           0.108844    0.428999   
..                                               ...         ...   
685                                         0.333333    0.214480   
686                                         0.455782    0.680617   
687                                         0.510204    0.548773   
688                                         0.551020    0.720172   
689                                         0.537415    0.129871   

     Peak2peak tiempo  TIEMPO peak2rms  ...  solidity    std_dev  \
0            0.164061         0.110557  ...         1   0.410795   
1            1.000000         0.238391  ...         1  12.640444   
2            0.154656         0.130988  ...         1   4.451864   
3            0.160484         0.092834  ...         1   7.681243   
4            0.379571         0.141011  ...         1   2.908373   
..                ...              ...  ...       ...        ...   
685          0.163796         0.131218  ...         1   2.546444   
686          0.504337         0.150002  ...         1  15.365171   
687          0.259238         0.084483  ...         1   9.442264   
688          0.407933         0.102759  ...         1  14.023466   
689          0.143861         0.179793  ...         1   0.485765   

     x_center_mass  x_centroid  x_centroid_cm_diff  y_center_mass  y_centroid  \
0         0.406519    0.405904           -0.000615       0.472183    0.471469   
1         0.240467    0.239852           -0.000615       0.279601    0.278887   
2         0.422509    0.421894           -0.000615       0.490728    0.490014   
3         0.418819    0.418204           -0.000615       0.486448    0.485735   
4         0.287208    0.286593           -0.000615       0.333809    0.333096   
..             ...         ...                 ...            ...         ...   
685       0.421894    0.421279           -0.000615       0.490014    0.489301   
686       0.325953    0.325338           -0.000615       0.378745    0.378031   
687       0.306888    0.306273           -0.000615       0.356633    0.355920   
688       0.337638    0.337023           -0.000615       0.392297    0.391583   
689       0.345633    0.345018           -0.000615

In [20]:

# Extraer la columna "Etiqueta"
etiquetas = df['Etiqueta']

# Normalizar las columnas numéricas
cols_numericas = df.columns[:-3]  # Excluyendo las columnas "Etiqueta" y "espectrograma etiqueta"
scaler = MinMaxScaler()
df[cols_numericas] = scaler.fit_transform(df[cols_numericas])

# Unir la columna "Etiqueta" a los datos normalizados
df_normalized = pd.concat([df[cols_numericas], etiquetas], axis=1)

# Mostrar el resultado
df_normalized

Media tiempo  Desviacion estandar tiempo  Varianza tiempo  \
0        0.099520                    0.226593         0.102626   
1        0.085819                    0.873569         0.795442   
2        0.077419                    0.211654         0.093624   
3        0.102426                    0.273302         0.132812   
4        0.091737                    0.414361         0.242705   
..            ...                         ...              ...   
685      0.101867                    0.194363         0.083598   
686      0.086898                    0.672442         0.516633   
687      0.077123                    0.537215         0.361351   
688      0.079244                    0.713007         0.568258   
689      0.087387                    0.107590         0.039672   

     Entropia tiempo  kurtosis tiempo  MSE tiempo  \
0                0.0         0.013447    0.907095   
1                0.0         0.050793    0.351758   
2                0.0         0.015484    0.882481   
3                0.0         0.011577    0.910015   
4                0.0         0.026587    0.723575   
..               ...              ...         ...   
685              0.0         0.018980    0.875568   
686              0.0         0.038755    0.298080   
687              0.0         0.024871    0.207060   
688              0.0         0.031290    0.165912   
689              0.0         0.010445    0.986443   

     Tiempo en alcanzar el pico maximo en frecuencia  RMS tiempo  \
0                                           0.115646    0.226602   
1                                           0.306122    0.873538   
2                                           0.122449    0.211665   
3                                           0.102041    0.273296   
4                                           0.108844    0.414371   
..                                               ...         ...   
685                                         0.333333    0.194357   
686                                         0.455782    0.672435   
687                                         0.510204    0.537213   
688                                         0.551020    0.713004   
689                                         0.537415    0.107580   

     Peak2peak tiempo  TIEMPO peak2rms  ...  perimeter  skewness  solidity  \
0            0.163998         0.110557  ...   0.750000  0.262347       0.0   
1            1.000000         0.238391  ...   0.402174  0.009667       0.0   
2            0.154592         0.130988  ...   0.827640  0.046256       0.0   
3            0.160420         0.092834  ...   0.874224  0.024948       0.0   
4            0.379524         0.141011  ...   0.647516  0.071298       0.0   
..                ...              ...  ...        ...       ...       ...   
685          0.163733         0.131218  ...   0.354037  0.062165       0.0   
686          0.504299         0.150002  ...   0.583851  0.012588       0.0   
687          0.259182         0.084483  ...   0.591615  0.020329       0.0   
688          0.407889         0.102759  ...   0.482919  0.014551       0.0   
689          0.143796         0.179793  ...   0.462733  0.203596       0.0   

      std_dev  x_center_mass  x_centroid  x_centroid_cm_diff  y_center_mass  \
0    0.014767       0.503906    0.503906                 0.0       0.503906   
1    0.499546       0.152344    0.152344                 0.0       0.152344   
2    0.174953       0.537760    0.537760                 0.0       0.537760   
3    0.302965       0.529948    0.529948                 0.0       0.529948   
4    0.113770       0.251302    0.251302                 0.0       0.251302   
..        ...            ...         ...                 ...            ...   
685  0.099423       0.536458    0.536458                 0.0       0.536458   
686  0.607553       0.333333    0.333333                 0.0       0.333333   
687  0.372771       0.292969    0.292969                 0.0       0.292969   
688  0.554368       0.358073   

In [21]:
corr = df_normalized.corr()
corr

<ipython-input-21-5033528e5490>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  corr = df_normalized.corr()


Media tiempo  Desviacion estandar tiempo  \
Media tiempo                    1.000000                    0.017338   
Desviacion estandar tiempo      0.017338                    1.000000   
Varianza tiempo                 0.002584                    0.984670   
Entropia tiempo                      NaN                         NaN   
kurtosis tiempo                 0.017619                   -0.000912   
...                                  ...                         ...   
x_center_mass                   0.035696                   -0.396811   
x_centroid                      0.035696                   -0.396811   
x_centroid_cm_diff                   NaN                         NaN   
y_center_mass                   0.035756                   -0.396924   
y_centroid                      0.035756                   -0.396924   

                            Varianza tiempo  Entropia tiempo  kurtosis tiempo  \
Media tiempo                       0.002584              NaN         0.017619   
Desviacion estandar tiempo         0.984670              NaN        -0.000912   
Varianza tiempo                    1.000000              NaN        -0.000600   
Entropia tiempo                         NaN              NaN              NaN   
kurtosis tiempo                   -0.000600              NaN         1.000000   
...                                     ...              ...              ...   
x_center_mass                     -0.398792              NaN        -0.130066   
x_centroid                        -0.398792              NaN        -0.130066   
x_centroid_cm_diff                      NaN              NaN              NaN   
y_center_mass                     -0.399071              NaN        -0.130580   
y_centroid                        -0.399071              NaN        -0.130580   

                            MSE tiempo  \
Media tiempo                 -0.027403   
Desviacion estandar tiempo   -0.659774   
Varianza tiempo              -0.638322   
Entropia tiempo                    NaN   
kurtosis tiempo              -0.230536   
...                                ...   
x_center_mass                 0.405632   
x_centroid                    0.405632   
x_centroid_cm_diff                 NaN   
y_center_mass                 0.406138   
y_centroid                    0.406138   

                            Tiempo en alcanzar el pico maximo en frecuencia  \
Media tiempo                                                      -0.054062   
Desviacion estandar tiempo                                        -0.155504   
Varianza tiempo                                                   -0.116200   
Entropia tiempo                                                         NaN   
kurtosis tiempo                                                    0.299245   
...                                                                     ...   
x_center_mass                                                     -0.346366   
x_centroid                                                        -0.346366   
x_centroid_cm_diff                                                      NaN   
y_center_mass                                                     -0.347876   
y_centroid                                                        -0.347876   

                            RMS tiempo  Peak2peak tiempo  TIEMPO peak2rms  \
Media tiempo                  0.021865          0.067038         0.038086   
Desviacion estandar tiempo    0.999987          0.760312        -0.219391   
Varianza tiempo               0.984626          0.710728        -0.230947   
Entropia tiempo                    NaN               NaN              NaN   
kurtosis tiempo              -0.000895          0.326775         0.756383   
...                                ...               ...              ...   
x_center_mass                -0.396629         -0.381176        -0.077526   
x_centroid                   -0.396629         -0.381176        -0.077526   
x_centroid_cm_diff                 NaN     

In [ ]:
#Pregunta 2 , 3 , 4
#Gain ratio - funcion objetiva
def objective_function(selected_indices):
    # Calculate Gain Ratio for the selected features
    info_gain = mutual_info_classif(X_train.iloc[:, selected_indices], Y_train, discrete_features='auto')
    intrinsic_info = [entropy(X_train.iloc[:, i]) for i in selected_indices]

    # Calculate Gain Ratio as the fitness score
    gain_ratio = [ig / ii if ii > 0 else 0 for ig, ii in zip(info_gain, intrinsic_info)]

    # Sum the Gain Ratios to get the overall fitness score (maximize Gain Ratio)
    fitness = sum(gain_ratio)

    return -fitness  # Minimize the negative fitness (maximize Gain Ratio)

# Artificial Bee Colony algorithm
def artificial_bee_colony(objective_function, n_dim, n_iter):
    best_solution = None
    best_fitness = float('-inf')

    for _ in range(n_iter):
        # Generate random solution
        solution = np.random.randint(2, size=n_dim)

        # Evaluate solution
        fitness = objective_function(solution)

        # Update best solution
        if fitness > best_fitness:
            best_solution = solution
            best_fitness = fitness

    return best_solution

# Convertir etiquetas a números
le = LabelEncoder()
y_encoded = le.fit_transform(df_normalized['Etiqueta'])

# Dividir los datos en entrenamiento y prueba
X_train, X_test, Y_train, Y_test = train_test_split(df_normalized.drop('Etiqueta', axis=1), y_encoded, test_size=0.2, random_state=42)


# Perform feature selection using ABC
selected_features = artificial_bee_colony(objective_function, n_dim=X_train.shape[1], n_iter=100)


# Print selected features
selected_indices = np.where(selected_features)[0]
print("Selected Features:", selected_indices)


Selected Features: [  1   2   3   4   5   8   9  12  14  15  18  19  20  21  23  24  25  26
  27  29  31  32  33  35  36  37  39  40  44  49  53  55  56  57  58  60
  62  64  65  66  67  69  70  74  75  76  78  79  80  81  82  83  84  85
  86  87  88  90  91  92  94  97  98  99 103 104 106]


In [ ]:
import numpy as np

class FeatureSelectionABC:
    def __init__(self, data, labels, n_bees, max_cycles, limit):
        self.data = data
        self.labels = labels
        self.n_bees = n_bees
        self.max_cycles = max_cycles
        self.limit = limit
        self.dim = data.shape[1]  # Number of features
        self.population = self.initialize_population()
        self.fitness = np.zeros(n_bees)
        self.trial_counter = np.zeros(n_bees)
        self.best_solution = None
        self.best_fitness = -np.inf

    def initialize_population(self):
        return np.random.randint(2, size=(self.n_bees, self.dim))

    def evaluate(self, solution):
        return np.random.rand()

    def mutate_solution(self, solution):
        idx = np.random.randint(0, self.dim)
        new_solution = solution.copy()
        new_solution[idx] = 1 - new_solution[idx]
        return new_solution

    def update_bees(self):
        for i in range(self.n_bees):
            new_solution = self.mutate_solution(self.population[i])
            new_fitness = self.evaluate(new_solution)
            if new_fitness > self.fitness[i]:
                self.population[i] = new_solution
                self.fitness[i] = new_fitness
                self.trial_counter[i] = 0
            else:
                self.trial_counter[i] += 1

    def scout_bees(self):
        for i in range(self.n_bees):
            if self.trial_counter[i] > self.limit:
                self.population[i] = np.random.randint(2, size=self.dim)
                self.fitness[i] = self.evaluate(self.population[i])
                self.trial_counter[i] = 0

    def optimize(self):
        for _ in range(self.max_cycles):
            self.update_bees()
            self.scout_bees()
            max_fitness_idx = np.argmax(self.fitness)
            if self.fitness[max_fitness_idx] > self.best_fitness:
                self.best_solution = self.population[max_fitness_idx]
                self.best_fitness = self.fitness[max_fitness_idx]


In [22]:
X_normalized = scaler.fit_transform(df_normalized.drop(['Etiqueta'], axis=1))
y = df_normalized['Etiqueta']

# 'X_normalized' es el conjunto de datos normalizado y 'y' son las etiquetas
X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.2, random_state=42)

# Implementación de la función objetivo GR
def gain_ratio(X_subset, y):
    # Utiliza SelectKBest y mutual_info_classif para calcular el Gain Ratio
    selector = SelectKBest(score_func=mutual_info_classif, k='all')
    selector.fit(X_subset, y)
    scores = selector.scores_

    # Calcula el Gain Ratio (simplificado, ajusta según sea necesario)
    gain_ratio = np.mean(scores)

    return gain_ratio

# Implementación ABC
def artificial_bee_colony(X, y, num_iterations, num_employed_bees, num_onlooker_bees):
    num_features = X.shape[1]
    results = []

    for _ in range(num_iterations):
        # Randomly generate employed bees' solutions
        solutions = np.random.choice([0, 1], size=(num_employed_bees, num_features))

        # Evaluate the solutions
        scores = np.zeros(num_employed_bees)
        for i in range(num_employed_bees):
            selected_features = np.where(solutions[i] == 1)[0]
            X_subset = X[:, selected_features]
            score = gain_ratio(X_subset, y)  # Utiliza la función objetivo GR
            scores[i] = score

            # Guarda los resultados
            results.append((selected_features, score))

        # Update solutions based on scores
        for i in range(num_employed_bees):
            if np.random.rand() < scores[i] / np.max(scores):
                solutions[i] = np.random.choice([0, 1], size=num_features)

    # Ordena los resultados por importancia
    results.sort(key=lambda x: x[1], reverse=True)

    return results[:5]  # Devuelve los cinco mejores subconjuntos

# Ejemplo de ejecución
num_iterations = 50
num_employed_bees = 10
num_onlooker_bees = 10

top_five_subsets = artificial_bee_colony(X_train, y_train, num_iterations, num_employed_bees, num_onlooker_bees)

# Imprime los cinco mejores subconjuntos y su importancia per cápita
for i, (subset, importance) in enumerate(top_five_subsets):
    print(f"Top {i+1} Subset: {subset}, Importance: {importance}")

# Entrena un clasificador con el subconjunto seleccionado y evalúalo en el conjunto de prueba
best_subset = top_five_subsets[0][0]
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train[:, best_subset], y_train)
y_pred = clf.predict(X_test[:, best_subset])

# Calcula el AUC score
auc_score = accuracy_score(y_test, y_pred)
print("AUC Score:", auc_score)

Top 1 Subset: [  4   5   6   7   9  10  13  14  19  22  30  31  33  36  39  40  42  43
  45  46  47  49  51  52  54  56  57  58  59  63  64  69  74  75  76  78
  83  85  86  88  89  90  91  93  94  97  98  99 102 103], Importance: 0.0772409789652269
Top 2 Subset: [  2   3   6  10  16  17  20  21  22  24  25  26  28  29  30  31  34  37
  38  44  48  49  51  52  54  55  56  57  59  62  64  68  72  75  77  78
  79  84  87  89  91  92  94  95  97 102 104 105 106 107], Importance: 0.07585138680716644
Top 3 Subset: [  1   2   4   6  10  14  19  24  32  36  37  38  39  40  42  45  47  49
  50  51  54  60  62  64  65  67  68  69  74  75  76  77  78  81  83  85
  91  97  98  99 101 102 103 106 107], Importance: 0.07334432223586143
Top 4 Subset: [  2   3   6   8   9  12  16  19  20  21  24  27  29  30  33  37  38  39
  40  43  47  49  52  54  56  57  58  59  60  63  64  65  68  69  70  71
  72  74  77  79  81  82  83  84  85  86  88  89  91  99 100 101 103 104
 107], Importance: 0.07329112637239

#Interpretación:
Top 1 Subset: [  4   5   6   7   9  10  13  14  19  22  30  31  33  36  39  40  42  43
  45  46  47  49  51  52  54  56  57  58  59  63  64  69  74  75  76  78
  83  85  86  88  89  90  91  93  94  97  98  99 102 103],
  Importance: 0.0772409789652269

Este conjunto tiene la mayor importancia per cápita según la métrica Gain Ratio. Las características enumeradas son las más relevantes para clasificar las clases.

Top 2 Subset: [  2   3   6  10  16  17  20  21  22  24  25  26  28  29  30  31  34  37
  38  44  48  49  51  52  54  55  56  57  59  62  64  68  72  75  77  78
  79  84  87  89  91  92  94  95  97 102 104 105 106 107],
  Importance: 0.07585138680716644

Este conjunto es el segundo más importante según la métrica Gain Ratio.

Top 3
Subset: [  1   2   4   6  10  14  19  24  32  36  37  38  39  40  42  45  47  49
  50  51  54  60  62  64  65  67  68  69  74  75  76  77  78  81  83  85
  91  97  98  99 101 102 103 106 107],
  Importance: 0.07334432223586143

Este conjunto es el tercero más importante según la métrica Gain Ratio.

Top 4 Subset: [  2   3   6   8   9  12  16  19  20  21  24  27  29  30  33  37  38  39
  40  43  47  49  52  54  56  57  58  59  60  63  64  65  68  69  70  71
  72  74  77  79  81  82  83  84  85  86  88  89  91  99 100 101 103 104
 107],
 Importance: 0.0732911263723907

Este conjunto es el cuarto más importante según la métrica Gain Ratio.

Top 5 Subset: [  4   6   9  13  19  21  22  23  25  28  29  31  32  36  37  42  43  44
  45  47  48  49  50  52  54  56  57  58  60  61  62  64  65  66  72  73
  75  76  77  78  80  81  83  84  89  90  92  95  97  98  99 103 104 106
 107],
 Importance: 0.07288605531246636


Este conjunto es el quinto más importante según la métrica Gain Ratio.

#Interpretación del AUC Score:

El AUC Score es 0.9191176470588235, lo que indica un rendimiento bastante bueno del clasificador basado en el subconjunto seleccionado. El AUC Score mide la capacidad del modelo para discriminar entre las dos clases. Un valor cercano a 1 sugiere un mejor rendimiento del modelo.

# Literatura

## Algoritmo ABC

- Karaboga, Dervis, and Bahriye Basturk. "Artificial Bee Colony (ABC) Optimization Algorithm for Solving Constrained Optimization Problems." Erciyes University Engineering Faculty, Department of Computer Engineering, 2007. [Paper](https://sci-hub.se/https://doi.org/10.1007/978-3-540-72950-1_77)
- [Visualizacion en YT](https://www.youtube.com/watch?v=3qQr1eZwz5E)
- [Popular packet implementation de ABC](https://pypi.org/project/beecolpy/)

## Gain Ratio
Estudio de la funcion objetiva Gain Ratio
  - COMPARATIVE STUDY OF ATTRIBUTE SELECTION USING GAIN RATIO
AND CORRELATION BASED FEATURE SELECTION [Paper](https://www.csjournals.com/IJITKM/PDF%203-1/19.pdf)
<br>
Librerias y referencia de origen de implementacion del codigo
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_classif.html
- https://github.com/maha-prathamesh/Feature-Selection-Techniques/blob/main/- Feature%20Selection%20Based%20on%20Mutual%20Information%20(Entropy)%20Gain.ipynb
- https://goldinlocks.github.io/Information-Gain/